### PR TITLE
Send components-changed only when there are actual changes

### LINKED ISSUE #83: Send new component messages only when a component actually changes
Now it seems that the coordinator sends a `component` message to FBP Protocol subscribers every time it receives a discovery message. This causes constant _Loading_ messages in Flowhub:

screenshot 2017-02-08 at 19 43 08

It would be better to send these messages only when the component information _actually_ changes (new ports, whatever), not every ping from participant.

### PR DESCRIPTION
**Pull request by bergie**
fixes #83

### CODE REVIEW

**Comment by jonnor** [comment]
What remains to be done here?

**Comment by bergie** [comment]
Right now doesn't work because JSON definition is different every time for situations where single component has multiple instances (and hence different queue names)

**Comment by jonnor** [comment]
Yeah, that information should not be stored in the component/library definition. It should ignored in `_updateDefinition`, and if we need to persist it on discovery, should be handled by `Coordinator`. Possibly in the `participants` datastructure

**Comment by jonnor** [comment]
Same with `role` and `id`, should also be stripped/ignored in Library

File: src/library.coffee [hunk 1]
```diff
@@ -102,6 +102,21 @@ normalizeConfig = (config) ->
 
   return config
 
+# Remove instance-specific data like role and extra from library data
+cleanComponentDefinition = (discovered) ->
+  return discovered unless discovered?.definition
```
Comment (jonnor) [inline src/library.coffee]: whats this for?

Comment (bergie) [inline src/library.coffee]: Some tests were failing without that check, but not sure why

File: src/library.coffee [hunk 2]
```diff
@@ -122,21 +137,28 @@ class Library extends EventEmitter
...
+        @components[name] = discovered
+        names.push name if names.indexOf(name) is -1
+        continue
+      unless JSON.stringify(existing.definition) is JSON.stringify(discovered.definition)
+        # updated
+        @components[name] = discovered
+        names.push name if names.indexOf(name) is -1
         continue
-      # update
-      for k, v of comp
```
Comment (jonnor) [inline src/library.coffee]: Doesn't this destroy `command` ? which is needed for instantiating component instances?

**Comment by jonnor** [comment]
Can you add checks in the tests? These are the two relevant places, I think
https://github.com/msgflo/msgflo/blob/master/spec/coordinator.coffee#L81
https://github.com/msgflo/msgflo/blob/master/spec/coordinator.coffee#L98

File: src/library.coffee (hunk 2 above)
Comment (bergie) [inline src/library.coffee]: Fair point, will fix

Comment (bergie) [inline src/library.coffee]: 211fe6c

**Comment by bergie** [comment]
@jonnor tests added

File: src/library.coffee [hunk 3]
```diff
@@ -122,21 +137,29 @@ class Library extends EventEmitter
...
+        for k, v of discovered
+          @components[name][k] = v
+        names.push name if names.indexOf(name) is -1
         continue
-      # update
-      for k, v of comp
-        existing[k] = v
 
+    # Skip sending components-changed if nothing changed
+    return unless names.length
```
Comment (jonnor) [inline src/library.coffee]:
```
if names.length
  @emit 'components-changed', names, @components
```
is much clearer than a conditional return (only really acceptable for failed preconditions, imo)

Comment (bergie) [inline src/library.coffee]: OK, handled in 3b85a8c

**Comment by bergie** [comment]
Finally getting some Travis results here

**Comment by jonnor** [comment]
Thanks!

**Pull request merged**